### DISCUSSION TITLE
Comparing only samples of specific classes

### DISCUSSION THREAD

**Discussion by danielcrane** [KevinMusgrave/pytorch-metric-learning#341] [Q&A] [answered]
Of the examples I've checked so far, they all involve creating models that compare all classes of samples from the dataset to one another, however I have a slightly different use-case that arises as a result of my current dataset structure.

In my case, I have a dataset of objects that looks something like so (with each `class_n` being a folder of samples of that class):

```
scene_0
---- class_0
---- class_1
---- class_2
scene_1
---- class_0
---- class_1
scene_2
---- class_0
---- class_1
---- class_2
---- class_3
```

and I'd like to compare only classes from within the same scene during training. The reason for this being that I cannot guarantee that each of the classes across scenes corresponds to the same thing, i.e. `scene_0/class_0` does not necessarily contain the same type of objects as `scene_1/class_0`.

In fact, it's incredibly unlikely that they're the same thing, but technically possible, so I'd like to avoid cross-`scene` comparisons as much as possible to avoid introducing noise.

As a complete novice to metric learning, I was wondering if this type of problem has a name (as I've been unable to find anything so far), and if anyone could provide advice on where I should start looking in order to modify for example the `MetricLossOnly.ipynb` notebook to work with this kind of problem. 
I originally thought that doing it at batch creation time might be the easiest thing, but I'm also concerned that having batches each with only classes from one `scene` might cause other issues during training.

**Comment by KevinMusgrave** [comment, marked as answer]
If ```scene_0/class_0``` doesn't share any characteristics of ```scene_1/class_0```, how about just renumbering the classes?

You could do this in your Dataset class, so you don't have to change the folders.

What I mean is something like this:

```
scene_0
---- class_0   --> 0
---- class_1   --> 1
---- class_2   --> 2
scene_1
---- class_0   --> 3
---- class_1   --> 4
scene_2
---- class_0   --> 5
---- class_1   --> 6
---- class_2   --> 7
---- class_3   --> 8
```

**Reply by danielcrane** [reply]
Thanks for your response @KevinMusgrave !

Unfortunately I can't guarantee that `scene_0/class_0` doesn't share any characteristics with `scene_1/class_0`, it's possible that they're the same thing, but also possible that they're not. 

The only information I have is about the distinctions between each object within a given `scene`.

It's a bit of a weird problem, and actually my dataset is a bit more complex than this, but if I were to do the renaming of classes as you mentioned, the problem would boil down to: I only want to compare certain classes to each other, but not all.

For example after the renumbering of classes, I'd only want to compare classes within the following tuples: (`0, 1, 2`), (`3, 4`), (`5, 6, 7, 8`).

I suppose the easiest way to do something like this would be making a custom data loader of some sorts, since if I were to wait until the mining stage it could be that the number of pairs that can be compared are too few form a meaningful batch.

**Reply by KevinMusgrave** [reply]
Here's an approach that might work. It assumes your labels are 2-dimensional, where the first dim is the class label and the second dim is the scene label. (Warning: I haven't tested this code)

```python
from pytorch_metric_learning.utils import loss_and_miner_utils as lmu

for data, labels in dataloader:
    class_labels = labels[:,0]
    scene_labels = labels[:,1]

    embeddings = model(data)
    loss = 0
    # collect all pairs per scene
    for s in torch.unique(scene_labels):
        mask = scene_labels == s
        curr_embeddings = embeddings[mask]
        curr_class_labels = class_labels[mask]
        curr_pairs = lmu.get_all_pairs_indices(curr_class_labels)
        loss += loss_fn(curr_embeddings, curr_class_labels, curr_pairs)

    loss.backward()
```

The only downside of this method is that you can't use a miner. However, you can use a threshold reducer in the loss function, which has a similar effect as a miner.

**Reply by danielcrane** [reply]
@KevinMusgrave Thanks so much, this is a really helpful starting point.

I'm a little surprised that this problem isn't more common, but I couldn't find anyone else with a similar use-case online at all!

**Reply by KevinMusgrave** [reply]
I guess "hierarchical metric learning" could be related

**Reply by danielcrane** [reply]
I'll look into it, thanks again!

I love how well documented this repo is by the way, really amazing work!

**Reply by KevinMusgrave** [reply]
Thanks! 😄
